### ISSUE TITLE
Intermittent `typeset -T` failure, possible stack handling bug

### ISSUE THREAD

**Issue by McDutchie** [ksh93/ksh#400]
The last line (130) in the following regression test intermittently fails on some systems: https://github.com/ksh93/ksh/blob/a7ed5d9fd76fb5387524ce953917c294cc6d0bee/src/cmd/ksh93/tests/types.sh#L114-L130

The failure occurs more often (consistently on the GitHub runner) after backporting the 93v- version of the src/lib/libast/regex code (see ffe84ee7898820948e89dc3739384765ea314ecd), which makes the regex code use a separate stack. This was reason for me to temporarily revert the separate stack aspect of the regex code (see b936c94fbcd9613687c9ce1e8139a03f7b5e74d2).

Now, after submitting this issue, I will re-backport that code and turn the failure into a warning (i.e. it is signalled but not counted as an error) so that the CI status does not stay on "failing". There's nothing wrong with the new regex code, but the fact that its use of a separate stack exposes a types bug possibly means the types code has some stack handling problem. This issue is for tracing and fixing that.

**Comment by McDutchie** [comment]
Now that's very interesting: all of a sudden, in the last few commits, the failure (now a warning) does not show up anymore. Whereas before, it was consistent on the github runner, after those regex changes.

I can't get it to fail/warn on Void Linux either, even when running the tests thousands of times.

Hmm. Between reverting those and reintroducing them, I did commit one fix for `typeset -T`, i.e., c5018f7c9523a1e9d2039ae83caf52a8011dec82. But I can't imagine how that could possibly be related, as it shouldn't affect anything at the execution level.

Since it looks like it might be gone, I guess I should turn the warning back into an actual error, so we don't fail to notice if it does show up again. I'll check the results of a few more CI runs for coming commits first.

**Comment by McDutchie** [comment]
So, this just decided to show up again on my macOS 10.14.6 x86_64 system:
```
test types begins at 2022-07-12+04:17:17
	types.sh[130]: FAIL: z.r.s should be z.r.x (expected foo, got '')
test types failed at 2022-07-12+04:17:18 with exit code 1 [ 99 tests 1 error ]
```

**Comment by hyenias** [comment]
On my macOS 10.13.6 x86_64 box with current, I also get errors very rarely, 3 out of 1000 iterations of types.sh (POSIX, C.UTF-8, shcomp) failed.
```
$ arch/*/bin/ksh --version
  version         sh (AT&T Research) 93u+m/1.1.0-alpha+c05d83aa 2022-07-14
$ for (( i=1001; --i; )); do bin/shtests types > /dev/null && echo $i: pass || echo $i: fail; done
1000: pass
...
	types.sh[555]: FAIL: typeset -p of nested type not correct (expected 'typeset -C b=(typeset -C -a ca=( [4]=(b_t b=(a_t b=(a=hello))));)', got $'typeset -C b=(typeset -C -a ca=( [4]=(b_t b=(a_t b=(a=\'\'))));)')
855: fail
...
	types.sh[128]: FAIL: z.r.x should be foo (got '')
	types.sh[130]: FAIL: z.r.s should be z.r.x (expected '', got foo)
725: fail
...
	types.sh[128]: FAIL: z.r.x should be foo (got '')
	types.sh[130]: FAIL: z.r.s should be z.r.x (expected '', got foo)
49: fail
48: pass
...
$
```
```
# All 1000 utf8 only pass
$ for (( i=1001; --i; )); do bin/shtests --utf8 types > /dev/null && echo $i: pass || echo $i: fail; done
# But, 1 out of 1000 failed for posix only
$ for (( i=1001; --i; )); do bin/shtests --posix types > /dev/null && echo $i: pass || echo $i: fail; done
...
	types.sh[128]: FAIL: z.r.x should be foo (got '')
	types.sh[130]: FAIL: z.r.s should be z.r.x (expected '', got foo)
184: fail
# All 1000 shcomp only pass
$ for (( i=1001; --i; )); do bin/shtests --compile types > /dev/null && echo $i: pass || echo $i: fail; done
```

**Comment by McDutchie** [comment]
I can no longer reproduce this. @hyenias, can you?

**Comment by McDutchie** [comment]
Here is the regression test in an infinite loop with counter for quicker testing. When I was still able to reproduce this problem, this script failed and exited quite quickly.
<details>
<summary>issue400.sh</summary>

```sh
err_exit() { echo "$0: $@" >&2; exit 1; }
integer i
while :
do
print -n "\\r$((++i)) "
(
typeset -T X_t=(
	typeset x=foo y=bar
	typeset s=${_.x}
	create()
	{
		_.y=bam
	}
)
X_t x
[[ ${x.x} == foo ]] || err_exit 'x.x should be foo'
[[ ${x.y} == bam ]] || err_exit 'x.y should be bam'
[[ ${x.s} == ${x.x} ]] || err_exit 'x.s should be x.x'
typeset -T Y_t=( X_t r )
Y_t z
[[ ${z.r.x} == foo ]] || err_exit "z.r.x should be foo (got $(printf %q "${z.r.x}"))"
[[ ${z.r.y} == bam ]] || err_exit "z.r.y should be bam (got $(printf %q "${z.r.y}"))"
[[ ${z.r.s} == ${z.r.x} ]] || err_exit "z.r.s should be z.r.x (expected $(printf %q "${z.r.x}"), got $(printf %q "${z.r.s}"))"
) || break
done
```
</details>

**Comment by hyenias** [comment]
Using Version AJM 93u+m/1.0.5-alpha 2023-02-20, your infinite loop regression test for the typeset -T failure ran without error for 100,000+ iterations before CTRL+C. I was entranced at seeing the iterations pass by so fast.

But since I was unsure that I was running the script correctly to test all 3 tests (posix, utf8, and shcomp), I ran what I did back when. Manual run of the types tests ran 1000 times without error. All good now. :)
